[TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42495 [TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs.**
* #42494 [TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs.
* #42493 [TensorExpr] Properly handle all dtypes in evaluation of CompareSelect exprs.

Differential Revision: [D22910753](https://our.internmc.facebook.com/intern/diff/D22910753)

Commit:
[TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs."


Differential Revision: [D22910753](https://our.internmc.facebook.com/intern/diff/D22910753)

[ghstack-poisoned]

nickgg commented on torch/csrc/jit/tensorexpr/eval.h:
```diff
@@ -574,7 +574,19 @@ class SimpleIREvaluator : public CodeGen, public IRVisitor {
+  case ScalarType::Name: {      \
+    cond_v = value_.as<Type>(); \
+  } break;
+      AT_FORALL_SCALAR_TYPES_AND(Bool, TYPE_CASE);
```
Not supporting Half? or just doesn't support coerce to bool?

Maybe a comment here either way.

ZolotukhinM replied:
I added an explicit case to throw a more verbose error for `Half`.

nickgg reviewed (approved):
💥

Commit:
Update on "[TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs."


Differential Revision: [D22910753](https://our.internmc.facebook.com/intern/diff/D22910753)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs."


Differential Revision: [D22910753](https://our.internmc.facebook.com/intern/diff/D22910753)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs."


Differential Revision: [D22910753](https://our.internmc.facebook.com/intern/diff/D22910753)

[ghstack-poisoned]
